### PR TITLE
#34:Upload docker image for main and for release and latest

### LINKED ISSUE #34: Fix docker image labels for merges to main and for releases
## Background

- we currently update the docker image with the label latest for each merge/push to main, but it would be better to have it point to the last release.

## Acceptance Criteria

We need a CI task
- publish-main, which builds the image whenever main get updated (merge, push), it then builds the docker image for the main branch and pushes it with the label main.
- publish-release, which builds the image whenever we create a release (release tag gets created), it then builds the docker image for the commit of this release and pushes it with the release name as label to docker hub. Furthermore, it advances the latest tag to release tag and builds the image also with the label `latest` and pushes this as well to docker hub.

`bash starter_scripts/build_docker_runner_image.sh <label>`
`bash starter_scripts/push_docker_runner_image.sh <label>`

### PR DESCRIPTION
**Pull request by tomuben**
Currently Docker images were submitted only for latest, the head of main branch.
1. We rename latest to main.
2. Whenever a new tag is created we submit one image with label latest and one image with the tag name.

Fixes #34

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      - name: Build new Docker image
+        run: "bash starter_scripts/build_docker_runner_image.sh"
+      - name: Docker login
+        run: echo "$SECRET_DOCKER_TOKEN" | docker login --username "$SECRET_DOCKER_USER_NAME" --password-stdin
+        env: # Set the secret as an input
+          SECRET_DOCKER_USER_NAME: ${{ secrets.DOCKER_USER_NAME }}
+          SECRET_DOCKER_TOKEN: ${{ secrets.DOCKER_TOKEN }}
+      - name: Push new Docker Release image
+        env:
+          RELEASE_TAG: ${{ github.event.release.tag_name }} #Details here https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#release
```
Comment (tkilias) [inline .github/workflows/release.yml:23]: you can use ${{ github.event.release.tag_name }} directly in the run definition

**Review by tkilias — CHANGES_REQUESTED** [review]

File: starter_scripts/push_docker_runner_image.sh [hunk 2]
```diff
@@ -3,9 +3,9 @@
 set -euo pipefail
 
 if [ -n "${1-}" ]; then
-  image_suffix="$1"
```
Comment (tkilias) [inline starter_scripts/push_docker_runner_image.sh:6]: Please add a comment that we allow multiple suffixes, and we build only ones and retag mutliple times

**Review by tkilias — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      - uses: actions/checkout@v2
+      - name: Build new Docker image
+        run: "bash starter_scripts/build_docker_runner_image.sh"
+      - name: Docker login
+        run: echo "$SECRET_DOCKER_TOKEN" | docker login --username "$SECRET_DOCKER_USER_NAME" --password-stdin
+        env: # Set the secret as an input
+          SECRET_DOCKER_USER_NAME: ${{ secrets.DOCKER_USER_NAME }}
+          SECRET_DOCKER_TOKEN: ${{ secrets.DOCKER_TOKEN }}
+      - name: Push new Docker images
+        run: "bash starter_scripts/push_docker_runner_image.sh latest ${{ github.event.release.tag_name }}"
```
Comment (tkilias) [inline .github/workflows/release.yml:22]: Quotes for ${{ github.event.release.tag_name }}

**Review by tkilias — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml (hunk 3 above)
Comment (tomuben) [reply .github/workflows/release.yml:22]: The whole string is already encapsulated by quotes....

**Review by tkilias — APPROVED** [review]

**Pull request merged**